File: 12-unit-testing-basics/src/main/java/org/first/tutorial/Exercise.java
package org.first.tutorial;

public class Exercise {

    /**
     * Simple battery helper to test.
     */
    public static class BatteryMonitor {
        private final double minVoltage;
        private final double maxVoltage;

        public BatteryMonitor(double minVoltage, double maxVoltage) {
            this.minVoltage = minVoltage;
            this.maxVoltage = maxVoltage;
        }

        public boolean isCritical(double voltage) {
            return voltage <= minVoltage;
        }

        public boolean isHealthy(double voltage) {
            return voltage >= maxVoltage;
        }

        public double percentRemaining(double voltage) {
            double clamped = Math.max(minVoltage, Math.min(voltage, maxVoltage));
            double range = maxVoltage - minVoltage;
            double remaining = clamped - minVoltage;
            return (remaining / range) * 100.0;
        }
    }
}

File: kotlin/12-unit-testing-basics/src/main/kotlin/org/first/tutorial/Exercise.kt
package org.first.tutorial

class Exercise {

    /**
     * Simple battery helper to test.
     */
    class BatteryMonitor(private val minVoltage: Double, private val maxVoltage: Double) {
        fun isCritical(voltage: Double): Boolean {
            return voltage <= minVoltage
        }

        fun isHealthy(voltage: Double): Boolean {
            return voltage >= maxVoltage
        }

        fun percentRemaining(voltage: Double): Double {
            val clamped = minOf(maxOf(voltage, minVoltage), maxVoltage)
            val range = maxVoltage - minVoltage
            val remaining = clamped - minVoltage
            return (remaining / range) * 100.0
        }
    }
}
